File: sdk/dotnet/src/FormaLib/ActionResult.cs
namespace Forma;

/// <summary>
/// Structured handler result — the wire form of the engine's ExecuteResult.
/// Handlers may also return plain data, which becomes <c>data</c>.
/// </summary>
public sealed class ActionResult
{
    private readonly List<Dictionary<string, object?>> _events = new();

    public ActionResult(object? data = null, string? newState = null)
    {
        Data = data;
        NewState = newState;
    }

    public object? Data { get; }
    public string? NewState { get; }

    /// <summary>Add an event emission to the result.</summary>
    public ActionResult WithEvent(string name, object? payload = null, bool durable = false)
    {
        var evt = new Dictionary<string, object?> { ["name"] = name };
        if (payload is not null)
        {
            // Convert anonymous objects to dictionaries for JSON serialization
            evt["payload"] = ToDictionary(payload);
        }
        if (durable) evt["durable"] = true;

        var result = new ActionResult(Data, NewState);
        result._events.AddRange(_events);
        result._events.Add(evt);
        return result;
    }

    /// <summary>Serialize to the wire format expected by forma-sidecar.</summary>
    internal Dictionary<string, object?> ToWire()
    {
        var wire = new Dictionary<string, object?> { ["data"] = Data };
        if (NewState is not null) wire["new_state"] = NewState;
        if (_events.Count > 0) wire["events"] = _events;
        return wire;
    }

    private static Dictionary<string, object?> ToDictionary(object obj)
    {
        var dict = new Dictionary<string, object?>();
        foreach (var prop in obj.GetType().GetProperties())
        {
            dict[prop.Name] = prop.GetValue(obj);
        }
        return dict;
    }
}

File: sdk/java/src/main/java/io/formspec/ActionResult.java
package io.formspec;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Structured handler result — the wire form of the engine's ExecuteResult.
 * Handlers may also return a plain Map/List, which becomes {@code data}.
 */
public final class ActionResult {
    private final Object data;
    private final String newState;
    private final List<Map<String, Object>> events;

    public ActionResult(Object data) {
        this(data, null);
    }

    public ActionResult(Object data, String newState) {
        this.data = data;
        this.newState = newState;
        this.events = new ArrayList<>();
    }

    /** Set the new state after this action. */
    public ActionResult withNewState(String newState) {
        return new ActionResult(this.data, newState);
    }

    /** Add an event emission to the result. */
    public ActionResult withEvent(String name) {
        return withEvent(name, Map.of(), false);
    }

    /** Add an event emission with a payload. */
    public ActionResult withEvent(String name, Map<String, Object> payload) {
        return withEvent(name, payload, false);
    }

    /** Add an event emission with a payload and durability flag. */
    public ActionResult withEvent(String name, Map<String, Object> payload, boolean durable) {
        var event = new LinkedHashMap<String, Object>();
        event.put("name", name);
        if (payload != null && !payload.isEmpty()) {
            event.put("payload", payload);
        }
        if (durable) {
            event.put("durable", true);
        }
        var result = new ActionResult(this.data, this.newState);
        result.events.addAll(this.events);
        result.events.add(event);
        return result;
    }

    /** Serialize to the wire format expected by formspec-sidecar. */
    public Map<String, Object> toWire() {
        var wire = new LinkedHashMap<String, Object>();
        wire.put("data", data);
        if (newState != null) {
            wire.put("new_state", newState);
        }
        if (!events.isEmpty()) {
            wire.put("events", events);
        }
        return wire;
    }
}
